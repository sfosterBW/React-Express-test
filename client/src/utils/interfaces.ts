export interface Fruit {
  id: string
  name: string
  best: boolean
  description?: string
}

export type NewFruit = Omit<Fruit, 'id'>
