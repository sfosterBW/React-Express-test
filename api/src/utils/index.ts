import { IFruit, NewFruit } from './types'

/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment*/

const isBoolean = (bool: unknown): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: unknown): text is string => {
  return typeof text === 'string' || text instanceof String;
}

const parseBest = (best: unknown): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`);
  }

  return best
}

const parseName = (name: unknown): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

const parseDescription = (description: unknown): string => {
  if (!description || !isString(description)) {
    throw new Error(`Incorrect or missing description: ${String(description)}`)
  }

  return description
}

export const toNewFruit = (object: any): NewFruit => {
  const newFruit: NewFruit = {
    name: parseName(object.name),
    best: parseBest(object.best),
    description: object.description ? parseDescription(object.description) : undefined
  }

  return newFruit
}

const parseId = (id: unknown): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

export const toFruit = (object: any): IFruit => {
  const fruit: IFruit = {
    ...object,
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best),
    description: object.description ? parseDescription(object.description) : undefined
  }

  return fruit
}
