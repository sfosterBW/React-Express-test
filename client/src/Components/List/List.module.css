.wrapper {
  margin: 1vmin 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  min-height: 30vh;
}

.titleWrapper {

}

.title {
  text-align: center;
  background: gray;
  height: 15vh;
}

.filters {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.filters button {
  display: flex;
  margin: 1vmin;
  padding: 1vmin;
  font-size: calc(10px + 1vmin);
  font-weight: bold;
  text-decoration: underline;
  border: none;
  border-radius: 0.5vmin;
  cursor: pointer;
}

.filter:hover {
  box-shadow: 0.2vmin 0.2vmin 0.5vmin 0.5vmin rgba(0, 0, 0, 0.25);
}

.devices {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.defaultText {
  color: black;
  font-size: 32px;
}
