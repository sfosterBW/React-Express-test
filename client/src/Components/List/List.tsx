import React, { FC, useState } from 'react'
import { RootState } from '../../Reducers/store'
import { useSelector } from 'react-redux'
import FruitItem from '../FruitItem/FruitItem'
import styles from './List.module.css'

const List: FC = () => {
  const [filter, setFilter] = useState<boolean | undefined>(undefined)
  const fruits = useSelector((state: RootState) => state.fruit
    .filter((i) => filter === i.best || filter === undefined)
  )
  const filteredList = () => {

    if (!fruits || fruits.length < 1 ) {
      return <p>Nothing here...</p>
    }

    return fruits.map(fruit => <FruitItem key={fruit.id} fruit={fruit} />)
  }

  const setter = (value: boolean | undefined) => (): void => setFilter(value)

  return (
    <section className={styles.wrapper}>
      <div className={styles.filters}>
        <button
          className={styles.filter}
          data-testid="all-filter"
          onClick={setter(undefined)}
        >
          All
        </button>
        <button
          className={styles.filter}
          data-testid="best-filter"
          onClick={setter(true)}
        >
          Best
        </button>
        <button
          className={styles.filter}
          data-testid="not-best-filter"
          onClick={setter(false)}
        >
          Not best
        </button>
      </div>
      <div className={styles.devices}>
        {filteredList()}
      </div>
    </section>
  )
}

export default List
