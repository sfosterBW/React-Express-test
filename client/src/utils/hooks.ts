import { useState } from 'react'

export const useField = (
  label: string,
  name: string,
  type: string,
  initValue: string = ""
) => {

  const [value, setValue] = useState(initValue)

  const handleChange = (event: any): void => {
    if (type === "checkbox") {
      setValue(event.target.checked)
    } else {
      setValue(event.target.value)
    }
  }

  const reset = (resetValue = ""): void => {
    setValue(resetValue)
  }

  return {
    handleChange,
    label,
    name,
    reset,
    type,
    value
  }
}
