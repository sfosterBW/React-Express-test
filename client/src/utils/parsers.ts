import { Fruit } from './interfaces'

/* eslint-disable  @typescript-eslint/no-explicit-any */
/* eslint-disable  @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */

const isBoolean = (bool: unknown): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: unknown): text is string => {
  return typeof text === 'string' || text instanceof String
}

const parseBest = (best: unknown): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`)
  }

  return best
}

export const parseId = (id: unknown): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

const parseName = (name: unknown): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

const parseDescription = (description: unknown): string => {
  if (!description || !isString(description)) {
    throw new Error(`Incorrect or missing description: ${String(description)}`)
  }

  return description
}

const parseArray = (array: any): Array<unknown> => {
  if (!array || !Array.isArray(array)) {
    throw new Error(`Incorrect or missing an array ${String(array)}`)
  }

  return array
}

export const toFruit = (object: any): Fruit => {
  const fruit: Fruit = {
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best)
  }

  if (object.description) {
    fruit.description = parseDescription(object.description)
  }

  return fruit
}

export const toFruits = (object: any): Fruit[] => {
  const array = parseArray(object)
  return array.map(fruit => toFruit(fruit))
}

export const toId = (object: any): string => {
  return parseId(object)
}

export default { toFruit, toFruits, toId }
