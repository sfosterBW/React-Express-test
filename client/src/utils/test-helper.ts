import { Fruit } from './interfaces'

export const emptyList: Fruit[] = []

export const fruit: Fruit = {
  id: "1",
  name:"apple",
  best:false,
  description: "I don't like doctors"
}

export const list: Fruit[] = [
  {
    id: "1",
    name: "apple",
    best: false,
    description: "I don't like doctors"
  },
  {
    id: "2",
    name: "banana",
    best: true
  }
]

export default { emptyList, fruit, list }
