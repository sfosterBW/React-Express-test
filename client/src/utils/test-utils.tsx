import React, { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureStore from "redux-mock-store"
import thunk from "redux-thunk"

import { initState as initAlertState } from '../Reducers/alertReducer'
import { initState as initFruitState } from '../Reducers/fruitReducer'
import { initState as initModalState } from '../Reducers/modalReducer'
import { RootState } from '../Reducers/store'

export const rootInitialState = {
  alert: initAlertState,
  fruit: initFruitState,
  modal: initModalState
};

export const renderWithProviders = (
  ui: ReactNode,
  initialState: RootState = rootInitialState
) => {
  const store = configureStore([thunk])(initialState)
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    mockStore: store,
    getActionsTypes: () => store.getActions().map(a => a.type)
  }
}
