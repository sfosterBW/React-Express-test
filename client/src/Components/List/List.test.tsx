import React from 'react'
import { fireEvent } from '@testing-library/react'
import { renderWithProviders, rootInitialState } from 'utils/test-utils'

import { list } from '../../utils/test-helper'

import List from './List'

//https://www.rrebase.com/posts/full-guide-to-testing-javascript-react
//https://github.com/testing-library/react-testing-library#complex-example

describe('the List component', () => {
  describe('works with no fruit', () => {

    it('renders with the right structure', () => {
      const { queryAllByTestId, getByText } = renderWithProviders(<List />)
      expect(queryAllByTestId('item')).toHaveLength(0)
      expect(getByText("Nothing here...")).toBeInTheDocument()
    })
  })

  describe('works with fruit', () => {
    it('renders with the right structure', () => {
      const { getAllByTestId } = renderWithProviders(<List />, {
        ...rootInitialState,
        fruit: list
      })
      expect(getAllByTestId('item')).toHaveLength(list.length)
      getAllByTestId('item-title').forEach((title: any, i: number) =>
        expect(title).toHaveTextContent(list[i].name)
      )
      getAllByTestId('item-best').forEach((best: any, i: number) =>
        expect(best).toHaveTextContent(String(list[i].best))
      )
    })

    it('filtering changes the list length', () => {
      const { getByTestId, getAllByTestId } = renderWithProviders(<List />, {
        ...rootInitialState,
        fruit: list
      })
      const bestList = list.filter(i => i.best === true)
      const notBestList = list.filter(i => i.best === false)

      fireEvent.click(getByTestId('best-filter'))
      expect(getAllByTestId('item')).toHaveLength(bestList.length)

      fireEvent.click(getByTestId('not-best-filter'))
      expect(getAllByTestId('item')).toHaveLength(notBestList.length)

      fireEvent.click(getByTestId('all-filter'))
      expect(getAllByTestId('item')).toHaveLength(list.length)
    })
  })
})
