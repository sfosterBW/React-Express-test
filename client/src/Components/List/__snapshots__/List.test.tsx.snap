// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the List component works with fruit renders the same as last time 1`] = `
<div>
  <section
    class="wrapper"
  >
    <div
      class="filters"
    >
      <button
        class="filter"
        data-testid="all-filter"
      >
        All
      </button>
      <button
        class="filter"
        data-testid="best-filter"
      >
        Best
      </button>
      <button
        class="filter"
        data-testid="not-best-filter"
      >
        Not best
      </button>
    </div>
    <div
      class="devices"
    >
      <div
        class="row"
        data-testid="item"
      >
        <div
          class="header"
        >
          <div
            class="section"
          >
            <h3
              class="subtitle"
              data-testid="item-title"
            >
              apple
            </h3>
          </div>
        </div>
        <div
          class="active"
          style="height: 0px; transition: all 0.5s;"
        >
          <div
            class="sectionRow"
          >
            <h4
              class="subtitle"
            >
              Description
            </h4>
            <p
              class="description"
              data-testid="item-description"
            >
              "I don't like doctors"
            </p>
          </div>
          <div
            class="section"
          >
            <h4
              class="subtitle"
            >
              Best
            </h4>
            <input
              class="button"
              id="1"
              name="1"
              type="checkbox"
              value="1"
            />
            <label
              data-testid="item-best"
              for="1"
            >
              false
            </label>
          </div>
          <div
            class="section"
          >
            <h4
              class="subtitle"
            >
              Edit
            </h4>
            <button
              class="button"
              data-testid="update-button"
              name="modal"
            >
              Update
            </button>
            <button
              class="button"
              data-testid="remove-button"
              name="remove"
              value="1"
            >
              Remove
            </button>
          </div>
        </div>
      </div>
      <div
        class="row"
        data-testid="item"
      >
        <div
          class="header"
        >
          <div
            class="section"
          >
            <h3
              class="subtitle"
              data-testid="item-title"
            >
              banana
            </h3>
          </div>
        </div>
        <div
          class="active"
          style="height: 0px; transition: all 0.5s;"
        >
          <div
            class="sectionRow"
          >
            <h4
              class="subtitle"
            >
              Description
            </h4>
            <p
              class="description"
              data-testid="item-description"
            >
              No description added yet
            </p>
          </div>
          <div
            class="section"
          >
            <h4
              class="subtitle"
            >
              Best
            </h4>
            <input
              checked=""
              class="button"
              id="2"
              name="2"
              type="checkbox"
              value="2"
            />
            <label
              data-testid="item-best"
              for="2"
            >
              true
            </label>
          </div>
          <div
            class="section"
          >
            <h4
              class="subtitle"
            >
              Edit
            </h4>
            <button
              class="button"
              data-testid="update-button"
              name="modal"
            >
              Update
            </button>
            <button
              class="button"
              data-testid="remove-button"
              name="remove"
              value="2"
            >
              Remove
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`the List component works with no fruit renders the same as last time 1`] = `
<div>
  <section
    class="wrapper"
  >
    <div
      class="filters"
    >
      <button
        class="filter"
        data-testid="all-filter"
      >
        All
      </button>
      <button
        class="filter"
        data-testid="best-filter"
      >
        Best
      </button>
      <button
        class="filter"
        data-testid="not-best-filter"
      >
        Not best
      </button>
    </div>
    <div
      class="devices"
    >
      <p>
        Nothing here...
      </p>
    </div>
  </section>
</div>
`;
